In [136]:
#import libraries
import pandas as pd
#import libraries
import sklearn
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
#import libraries
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [137]:
# Load the train_data.csv file in python, and assign it to a variable named "train".
##### YOUR CODE STARTS HERE #####
train = pd.read_csv('/content/train_data.csv')
##### YOUR CODE ENDS HERE #####
train

Gender  Age  Annual Income     Profession  Work Experience  Family Size  \
0       Male   91           9000     Healthcare                1            5   
1     Female   59         119833         Doctor                1            2   
2       Male   59          97185  Entertainment                0            5   
3     Female   39         158638     Healthcare                2            5   
4     Female   74         101352         Lawyer                0            5   
...      ...  ...            ...            ...              ...          ...   
1567  Female   54         119223         Lawyer               10            7   
1568  Female   34           8000  Entertainment                6            6   
1569    Male    6         107117         Artist                1            3   
1570  Female   75         135088         Lawyer                1            5   
1571  Female    3         136316      Executive                0            3   

      Spending Score Spending Category  
0                 96              High  
1                 38               Low  
2                 87              High  
3                 78              High  
4                 36               Low  
...              ...               ...  
1567              58              High  
1568              76              High  
1569              44               Low  
1570               5               Low  
1571              88              High  

[1572 rows x 8 columns]

In [138]:
# Load the test_data.csv file in python, and assign it to a variable named "test".
##### YOUR CODE STARTS HERE #####
test = pd.read_csv('/content/test_data.csv')
##### YOUR CODE ENDS HERE #####
test
# We upload the training and test file into pandas DataFrames. This allow us to work with the data in Python

Gender  Age  Annual Income     Profession  Work Experience  Family Size  \
0    Female   23          84000     Healthcare                1            3   
1    Female   35          93000     Healthcare                4            4   
2      Male   37          19000         Doctor                0            1   
3      Male   20          89000     Healthcare                1            6   
4      Male   52          20000  Entertainment                1            4   
..      ...  ...            ...            ...              ...          ...   
388  Female   41         128960      Executive                9            4   
389  Female   62         149797         Artist                5            6   
390    Male   94         181183      Marketing                9            3   
391    Male   77         182109      Executive                7            2   
392    Male   90         110610  Entertainment                5            2   

     Spending Score Spending Category  
0                94              High  
1                99              High  
2                13               Low  
3                66              High  
4                29               Low  
..              ...               ...  
388              43               Low  
389              19               Low  
390              24               Low  
391               4               Low  
392              52              High  

[393 rows x 8 columns]

In [139]:
# For more info about pd.get_dummies
# https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html

##### YOUR CODE STARTS HERE #####

# The categorical columns we want to encode
categorical_columns = ['Gender', 'Profession']

# Use pd.get_dummies() only on the specified columns
encoded_columns = pd.get_dummies(train[categorical_columns], drop_first=True)

# Concatenate the encoded columns with the original DataFrame (train)
train = pd.concat([train.drop(columns=categorical_columns), encoded_columns], axis=1)

# Drop the original categorical columns
train

##### YOUR CODE ENDS HERE #####

Age  Annual Income  Work Experience  Family Size  Spending Score  \
0      91           9000                1            5              96   
1      59         119833                1            2              38   
2      59          97185                0            5              87   
3      39         158638                2            5              78   
4      74         101352                0            5              36   
...   ...            ...              ...          ...             ...   
1567   54         119223               10            7              58   
1568   34           8000                6            6              76   
1569    6         107117                1            3              44   
1570   75         135088                1            5               5   
1571    3         136316                0            3              88   

     Spending Category  Gender_Male  Profession_Doctor  Profession_Engineer  \
0                 High         True              False                False   
1                  Low        False               True                False   
2                 High         True              False                False   
3                 High        False              False                False   
4                  Low        False              False                False   
...                ...          ...                ...                  ...   
1567              High        False              False                False   
1568              High        False              False                False   
1569               Low         True              False                False   
1570               Low        False              False                False   
1571              High        False              False                False   

      Profession_Entertainment  Profession_Executive  Profession_Healthcare  \
0                        False                 False                   True   
1                        False                 False                  False   
2                         True                 False                  False   
3                        False                 False                   True   
4                        False                 False                  False   
...                        ...                   ...                    ...   
1567                     False                 False                  False   
1568                      True                 False                  False   
1569                     False                 False                  False   
1570                     False                 False                  False   
1571                     False                  True                  False   

      Profession_Homemaker  Profession_Lawyer  Profession_Marketing  
0                    False              False                 False  
1                    False              False                 False  
2                    False              False                 False  
3                    False              False                 False  
4                    False               True                 False  
...                    ...                ...                   ...  
1567                 False               True                 False  
1568                 False              False                 False  
1569                 False              False                 False  
1570                 False               True                 False  
1571                 False              False                 False  

[1572 rows x 15 columns]

In [140]:
##### YOUR CODE STARTS HERE #####

# The categorical columns we want to encode
categorical_columns = ['Gender', 'Profession']

# Use pd.get_dummies() only on the specified columns
encoded_columns = pd.get_dummies(test[categorical_columns], drop_first=True)

# Concatenate the encoded columns with the original DataFrame (test)
test = pd.concat([test.drop(columns=categorical_columns), encoded_columns], axis=1)

# Drop the original categorical columns
test

##### YOUR CODE ENDS HERE #####
#We converted categorical columns ('Gender' and 'Profession') into numerical format using one-hot encoding.We then replaced the original categorical columns in both training and test datasets.and for # Drop the original categorical columns no action taken since the original categorical columns are already dropped during the concatenation step

Age  Annual Income  Work Experience  Family Size  Spending Score  \
0     23          84000                1            3              94   
1     35          93000                4            4              99   
2     37          19000                0            1              13   
3     20          89000                1            6              66   
4     52          20000                1            4              29   
..   ...            ...              ...          ...             ...   
388   41         128960                9            4              43   
389   62         149797                5            6              19   
390   94         181183                9            3              24   
391   77         182109                7            2               4   
392   90         110610                5            2              52   

    Spending Category  Gender_Male  Profession_Doctor  Profession_Engineer  \
0                High        False              False                False   
1                High        False              False                False   
2                 Low         True               True                False   
3                High         True              False                False   
4                 Low         True              False                False   
..                ...          ...                ...                  ...   
388               Low        False              False                False   
389               Low        False              False                False   
390               Low         True              False                False   
391               Low         True              False                False   
392              High         True              False                False   

     Profession_Entertainment  Profession_Executive  Profession_Healthcare  \
0                       False                 False                   True   
1                       False                 False                   True   
2                       False                 False                  False   
3                       False                 False                   True   
4                        True                 False                  False   
..                        ...                   ...                    ...   
388                     False                  True                  False   
389                     False                 False                  False   
390                     False                 False                  False   
391                     False                  True                  False   
392                      True                 False                  False   

     Profession_Homemaker  Profession_Lawyer  Profession_Marketing  
0                   False              False                 False  
1                   False              False                 False  
2                   False              False                 False  
3                   False              False                 False  
4                   False              False                 False  
..                    ...                ...                   ...  
388                 False              False                 False  
389                 False              False                 False  
390                 False              False                  True  
391                 False              False                 False  
392                 False              False                 False  

[393 rows x 15 columns]

In [141]:
#Drop Spending Category column
train_reg = train.drop(columns="Spending Category")
test_reg = test.drop(columns="Spending Category")

In [142]:
##### YOUR CODE STARTS HERE #####

# Prepare the training data (use train_reg set)
# Create the feature matrix X_train
# Exclude the target variable 'Spending Score'
X_train = train_reg.drop(columns=['Spending Score'])

# Extract the target variable y_train (use train_reg set)
# Assign the 'Spending Score' column to y_train
y_train = train_reg['Spending Score']

# Prepare the testing data (use test_reg set)
# Create the feature matrix X_test
# Exclude the target variable 'Spending Score'
X_test = test_reg.drop(columns=['Spending Score'])

# Extract the target variable y_test (use test_reg set)
# Assign the 'Spending Score' column to y_test
y_test = test_reg['Spending Score']

##### YOUR CODE ENDS HERE #####
#We separated the feature columns from the target column in both the training and test datasets.

In [143]:
# for more info about the model
# refer https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

##### YOUR CODE STARTS HERE #####

# Create an instance of the LinearRegression class
linear_reg_model = LinearRegression()

# Fit the linear regression model to the training data
linear_reg_model.fit(X_train, y_train)

# Use the trained linear regression model to make predictions on the test data
predictions = linear_reg_model.predict(X_test)

##### YOUR CODE ENDS HERE #####


In [144]:

##### YOUR CODE STARTS HERE #####

# Compute the MSE metric to evaluate the trained model using the test data
mse = mean_squared_error(y_test, predictions)

##### YOUR CODE ENDS HERE #####
print(mse)
#the Mean Squared Error to evaluate its performance.

780.7824887124046


In [145]:
#Drop Spending Category column
train_log_reg = train.drop(columns="Spending Score")
test_log_reg = test.drop(columns="Spending Score")

In [146]:
##### YOUR CODE STARTS HERE #####

# Prepare the training data (use train_log_reg set)
# Create the feature matrix X_train
# Exclude the target variable 'Spending Category'
X_train = train_log_reg.drop(columns="Spending Category")

# Extract the target variable y_train (use train_log_reg set)
# Assign the 'Spending Category' column to y_train
y_train = train_log_reg["Spending Category"]

# Prepare the testing data (use test_log_reg set)
# Create the feature matrix X_test
# Exclude the target variable 'Spending Category'
X_test = test_log_reg.drop(columns="Spending Category")

# Extract the target variable y_test (use test_log_reg set)
# Assign the 'Spending Category' column to y_test
y_test = test_log_reg["Spending Category"]

##### YOUR CODE ENDS HERE #####

In [147]:
# for more info about the model
# refer https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html
##### YOUR CODE STARTS HERE #####

# Create an instance of the Logistic regression class
logistic_reg_model = LogisticRegression()

# Fit the logistic regression model to the training data
logistic_reg_model.fit(X_train, y_train)

# Use the trained logistic regression model to make predictions on the test data
predictions = logistic_reg_model.predict(X_test)

##### YOUR CODE ENDS HERE #####


In [148]:
# for more info about the accuracy score
# refer https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html

##### YOUR CODE STARTS HERE #####

# Compute the Accuracy metric to evaluate the trained model on the test data
acc_metric = accuracy_score(y_test, predictions)

##### YOUR CODE ENDS HERE #####
print(acc_metric)
#the accuracy score to measure how many predictions were correct

0.455470737913486


**ANALYSIS** :The results from our models reflect a few key issues. The Linear Regression model produced a high Mean Squared Error (MSE) of 780.78, indicating that the model struggled to fit the data well, possibly due to data quality issues . Proper feature handling is crucial—while we converted categorical data correctly, missing or incorrect features can affect performance. The choice of models also mattered: Linear Regression may not have been the best fit for our data’s complexity, while the Logistic Regression model, with an accuracy of 0.455, struggled with classification. Tuning the models properly is essential, and it seems that we may not have optimized our model settings enough. To improve results, we need to focus on cleaning our data, selecting the right features, choosing appropriate models, and fine-tuning them more effectively.